### ISSUE TITLE
[Question] Residual Peeking in TopK and BatchTopK auxiliary losses

### ISSUE THREAD

**Issue by xXCoolinXx** [jbloomAus/SAELens#666]
# Questions

I had a question about the particular auxiliary loss implemented for the TopK and BatchTopK SAEs. 

The specific implementation used by this repo is 
```python3
def calculate_topk_aux_loss(
        self,
        sae_in: torch.Tensor,
        sae_out: torch.Tensor,
        hidden_pre: torch.Tensor,
        dead_neuron_mask: torch.Tensor | None,
    ) -> torch.Tensor:
        """
        Calculate TopK auxiliary loss.

        This auxiliary loss encourages dead neurons to learn useful features by having
        them reconstruct the residual error from the live neurons. It's a key part of
        preventing neuron death in TopK SAEs.
        """
        # Mostly taken from https://github.com/EleutherAI/sae/blob/main/sae/sae.py, except without variance normalization
        # NOTE: checking the number of dead neurons will force a GPU sync, so performance can likely be improved here
        if dead_neuron_mask is None or (num_dead := int(dead_neuron_mask.sum())) == 0:
            return sae_out.new_tensor(0.0)

        if self.cfg.normalize_activations in ("constant_norm_rescale", "layer_norm"):
            raise ValueError(
                "TopK auxiliary loss does not support activation normalization "
                f"(normalize_activations={self.cfg.normalize_activations!r}). "
                "The aux loss reconstruction would be in normalized space while the "
                "residual is in the original space, producing incorrect gradients."
            )

        residual = (sae_in - sae_out).detach()

        # Heuristic from Appendix B.1 in the paper
        k_aux = sae_in.shape[-1] // 2

        # Reduce the scale of the loss if there are a small number of dead latents
        scale = min(num_dead / k_aux, 1.0)
        k_aux = min(k_aux, num_dead)

        auxk_acts = calculate_topk_aux_acts(
            k_aux=k_aux,
            hidden_pre=hidden_pre,
            dead_neuron_mask=dead_neuron_mask,
        )

        # Encourage the top ~50% of dead latents to predict the residual of the
        # top k living latents. Per the paper (Appendix A.2), the reconstruction
        # is ê = W_dec @ z (no bias), since b_dec is already in the residual.
        recons = act_times_W_dec(
            auxk_acts, self.W_dec, self.cfg.rescale_acts_by_decoder_norm
        )
        # Apply the same reshaping as decode() so recons matches the residual's shape
        recons = self.reshape_fn_out(recons, self.d_head)
        auxk_loss = (recons - residual).pow(2).sum(dim=-1).mean()
        return self.cfg.aux_loss_coefficient * scale * auxk_loss

    @override
    def process_state_dict_for_saving_inference(
        self, state_dict: dict[str, Any]
    ) -> None:
        super().process_state_dict_for_saving_inference(state_dict)
        if self.cfg.rescale_acts_by_decoder_norm:
            _fold_norm_topk(
                W_enc=state_dict["W_enc"],
                b_enc=state_dict["b_enc"],
                W_dec=state_dict["W_dec"],
            )
```

More simply, what this code is doing is selecting the top $k//2$ dead latent variables and providing a gradient signal based on reconstructing the residual. 

The issue is that, if for example a latent or group of latents is undertrained, it always produces a high residual. We want those original latents to be solely responsible for whatever feature they are modeling (linear direction, manifold, whatever), but it is possible that latents chosen by the auxiliary loss essentially become error correcting terms to reduce this high residual, making it so that neither the original latent(s) modeling the feature nor the error correcting term fully describe the feature on their own. 

My main question is if there is anything that could be done to fix this. I think one obvious solution is just using a JumpReLU SAE, since it relies on a pre-activation loss for not firing, rather than using an auxiliary loss based on the residual. 

But this is undesirable to me, for I've found that, once I make any form of architectural modification to the SAE, the JumpReLU hyperparameters need to be completely retuned (and sometimes the function itself needs to be modified). BatchTopK is nice in that it is flexible enough to be used with a much wider variety of architectural modifications without significant tuning. 

I'm curious if there might be some straight through estimator style approach for TopK/BatchTopK that can be used to selectively apply a small gradient signal that is more "natural" and doesn't lead to this error correction issue. And also whether anyone thinks this is an actual issue or if I'm misinterpreting something.

**Comment by chanind** [comment]
This seems like a reasonable concern to me, but I'm also not sure how to go about fixing it or telling if it really is occuring in LLM SAEs.

I've noticed that there are sometimes dead latent issues with BatchTopK SAEs with certain settings in [SynthSAEBench](https://decoderesearch.github.io/SAELens/latest/synth_sae_bench/) that might be due to what you're describing (I can't remember the exact settings that cause this, try training intentionally too high or too low L0 as a starting point, or bumping up firing probabilities a bunch). It's strange in that the dead latents aren't truly "dead", they fire but extremely rarely, much more rarely than they should based on the features being tracked. Then the aux loss tries to revive them, but this knocks out some other existing latents into this strange "pseudo-dead" space, where the latent fires only a super-specific combination of other latents (sort of like the [red-triangle problem](https://www.lesswrong.com/posts/QoR8noAB3Mp2KBA4B/do-sparse-autoencoders-find-true-features)).

I'm not sure if this is the failure mode you're talking about, but if you can get it to happen in a synthetic setting it might be a lot easier to experiment with different aux losses to address it. Or just tell claude to figure out a new aux loss that fixes the dead latent issues.

**Comment by xXCoolinXx** [comment]
I'm in the same boat as you, in that my empirical results are from a fairly different setting from a normal SAE, though I am training on LLMs. I can say with confidence that it seems to be an issue with the aux loss though, as I initially had aux loss broken in my research architecture, and the latent representations were very nice and interpretable. I fixed aux loss, but a lot of those nice and interpretable representations went away (though I stopped having like 1/4 of my SAE dead). I've tried using different L0 values and aux loss coefficients and it seems like this issue is pretty intrinsic to the aux loss. 

I think the issue you were seeing is basically the exact same issue I've been facing. That seems like essentially what I would expect to happen in a normal SAE if the above issue is real. It is essentially like the zombie latents are coming alive and stealing variance away from the healthy latents. 

I would imagine there could be some latent for detecting shape words, but it is a bit bad at detecting `square`. The aux loss comes online, and a dead latent comes in and starts predicting `square`, and later several other shapes. Suddenly the latent that was alive now only represents `circle`. I think this is essentially what is happening, but this could be too simplistic.

The dynamics in my architecture are somewhat different, but I have observed interpretable representations that were localized to one latent or group of latents becoming more spread out and less interpretable.

I think maybe the fix is actually quite easy, in just heisting the JumpReLU loss from Anthropic

<img width="346" height="55" alt="Image" src="https://github.com/user-attachments/assets/0b47996b-4d2c-4a55-b32d-d495088f5186" />
 
and changing out the $\exp(t)$ for the global threshold for the BatchTopK. The global threshold changes quite smoothly and doesn't have any gradient so I think it would be fine to use there. I will run it with my architecture and see what happens.

**Comment by JordanMcCann** [comment]
Ran a quick experiment to test this directly. Short version: I don't think AuxK creates the error-correctors you're worried about, but the concern isn't unreasonable - it's a coefficient problem, not a structural one. What AuxK actually does, since the framing matters: it only touches the dead latents, takes the top-k_aux of them, and has them reconstruct the residual `e = x − x̂` of the main top-k reconstruction. Small coefficient, ~1/32 in Gao et al. None of it enters the main forward pass - at inference you still only use the live top-k. So a dead latent isn't wired in as a standing error-corrector. AuxK is a training-time nudge to keep gradient flowing so it can climb back into the top-k.

The reframe I'd push: pointing a dead latent at the residual is pointing it at the part of the input nothing currently explains. That's exactly where the next real feature lives. The residual target isn't at odds with learning a genuine feature. Your worry only bites if a cluster of dead latents gets pushed to jointly mop up residual in a dense, non-sparse way - and that's a k_aux/coefficient-too-large failure, not an inherent one.

I tried to induce that failure in a tiny setup: TopK SAE on synthetic sparse data, deliberately under-capacity (k=8, 10 active ground-truth features, so there's real leftover residual). Measured whether AuxK-revived latents come back denser than normal ones:

  ```
  aux_coef | dead % |  FVU  | n_revived | revived fire | normal fire
     0.000 |  8.6%  | 0.299 |     0     |     -        |    0.068
     1/32  |  8.6%  | 0.297 |     2     |   0.028      |    0.069
     1/4   |  6.2%  | 0.295 |     4     |   0.029      |    0.068
     1.0   |  5.5%  | 0.287 |     5     |   0.051      |    0.067
  ```

AuxK revived a handful of dead latents, improved FVU slightly, and the revived ones fired less often than typical latents - sparser, not dense error-correctors - even at coefficient 1.0, which is 32× the usual. The opposite of the failure mode. Caveat: toy scale, planted linear features, no superposition, only a few revived. I wouldn't over-read it.

On the straight-through estimator: the instinct is right, but it targets a different problem. STE - JumpReLU, Rajamanoharan et al. 2024 - gives a main-loss gradient to latents near the threshold, so marginally-not-selected ones keep learning. But a fully-dead latent whose preactivations have collapsed is nowhere near the cutoff, so STE passes it almost no gradient either. You still want some revival mechanism for the truly-dead case. They're complementary: JumpReLU/STE for the marginal latents, a small revival term only for the hard-dead ones.

So: AuxK aims dead units at unexplained variance, which is where features are. The dense-memorization failure is real, but it's a coefficient problem, and it's watchable - track AuxK loss as a fraction of total, and whether revived latents come back dense or fail auto-interp.

  Repro's below. Tell me where I'm wrong.

  <details><summary>auxk_demo.py (torch, CPU, ~30s)</summary>

  ```python
  import torch, torch.nn.functional as F
  torch.set_num_threads(4)

  d_model, n_true, n_latents = 64, 160, 128
  k, k_aux = 8, 16
  dead_after = 150
  B, STEPS = 512, 3000

  torch.manual_seed(0)
  G = F.normalize(torch.randn(n_true, d_model), dim=1)

  def sample(batch):
      s = 10  # 10 active true features, but k=8 live latents -> real residual
      idx = torch.randint(0, n_true, (batch, s))
      coeff = torch.rand(batch, s) + 0.2
      a = torch.zeros(batch, n_true).scatter_(1, idx, coeff)
      return a @ G + 0.03 * torch.randn(batch, d_model)

  def train(aux_coeff):
      torch.manual_seed(1)
      E = torch.nn.Parameter(torch.randn(d_model, n_latents) * (d_model ** -0.5))
      D = torch.nn.Parameter(F.normalize(torch.randn(n_latents, d_model), dim=1).clone())
      b = torch.nn.Parameter(torch.zeros(n_latents))
      opt = torch.optim.Adam([E, D, b], lr=1e-3)
      steps_since_fired = torch.zeros(n_latents)
      ever_dead = torch.zeros(n_latents, dtype=torch.bool)
      for _ in range(STEPS):
          x = sample(B)
          pre = F.relu(x @ E + b)
          topv, topi = pre.topk(k, dim=1)
          z = torch.zeros_like(pre).scatter_(1, topi, topv)
          xhat = z @ F.normalize(D, dim=1)
          resid = x - xhat
          loss = (resid ** 2).mean()
          dead = steps_since_fired > dead_after
          ever_dead |= dead
          if aux_coeff > 0 and dead.any():
              pre_dead = pre.masked_fill(~dead.unsqueeze(0), 0.0)
              kk = min(k_aux, int(dead.sum()))
              av, ai = pre_dead.topk(kk, dim=1)
              z_aux = torch.zeros_like(pre).scatter_(1, ai, av)
              ehat = z_aux @ F.normalize(D, dim=1)
              loss = loss + aux_coeff * ((resid.detach() - ehat) ** 2).mean()
          opt.zero_grad(); loss.backward(); opt.step()
          fired = torch.zeros(n_latents, dtype=torch.bool).scatter_(0, topi.reshape(-1), True)
          steps_since_fired = torch.where(fired, torch.zeros_like(steps_since_fired), steps_since_fired + 1)
      with torch.no_grad():
          x = sample(8192)
          pre = F.relu(x @ E + b)
          topv, topi = pre.topk(k, dim=1)
          z = torch.zeros_like(pre).scatter_(1, topi, topv)
          xhat = z @ F.normalize(D, dim=1)
          fvu = ((x - xhat) ** 2).sum() / ((x - x.mean(0)) ** 2).sum()
          firing = torch.zeros(n_latents).scatter_add_(0, topi.reshape(-1), torch.ones(topi.numel())) / x.shape[0]
          ever_fired = firing > 0
          revived = ever_dead & ever_fired
          normal = (~ever_dead) & ever_fired
          rrate = float(firing[revived].mean()) if revived.any() else float("nan")
          nrate = float(firing[normal].mean()) if normal.any() else float("nan")
          return aux_coeff, float(1 - ever_fired.float().mean()), float(fvu), int(revived.sum()), rrate, nrate

  print(f"{'aux':>6} | {'dead%':>5} | {'FVU':>5} | {'n_rev':>5} | {'rev fire':>8} | {'norm fire':>9}")
  for a in [0.0, 1/32, 1/4, 1.0]:
      aux, d, fvu, nr, rr, nn = train(a)
      print(f"{aux:>6.3f} | {100*d:>4.1f}% | {fvu:>5.3f} | {nr:>5d} | {rr:>8.3f} | {nn:>9.3f}")
  ```
  </details>

**Comment by xXCoolinXx** [comment]
Thank you, Claude Opus 4.8... for a complete load of nonsense. Your analysis does not consider anything to do with interpretability metrics, and you don't even provide any analysis about whether the SAE found canonical features on a TOY MODEL. Use Fable 5 next time. 

Thank you for your attention to this matter.
